File: animations/black_hole.py
import math

from core import Animation, BLACK_BG, RESET, density_char, gray_fg, smoothstep, star_noise


BLACK_HOLE_CHAR_THRESHOLDS = [(0.10, " "), (0.18, "."), (0.28, ":"), (0.39, "-"), (0.51, "="), (0.64, "+"), (0.78, "*"), (0.91, "#"), (1.01, "%")]


def black_hole_char(level):
  return density_char(level, BLACK_HOLE_CHAR_THRESHOLDS)


def accretion_band(x, y, radius, thickness, t, density, boost):
  disk_radius = math.hypot(x, y)
  disk_theta = math.atan2(y, x)
  ring = math.exp(-((disk_radius - radius) ** 2) / thickness)
  azimuth = 0.62 + 0.38 * math.sin(9.0 * disk_theta - 2.2 * t)
  turbulence = 0.72 + 0.28 * abs(math.sin(31.0 * disk_radius * density + 5.0 * disk_theta - 3.4 * t))
  return ring * azimuth * turbulence * boost


class BlackHoleAnimation(Animation):
  def render(self, width, height, elapsed, phase, options):
    radius = max(1.0, min(width, height * 2) * 0.44)
    density = max(0.45, options.scale)
    t = elapsed * 0.55
    rows = []
    for row in range(height):
      line = [BLACK_BG]
      last_color = None
      py = ((row - (height - 1) * 0.5) * 2.0) / radius
      for col in range(width):
        px = (col - (width - 1) * 0.5) / radius
        r = math.hypot(px, py)
        lens = 0.18 / (r * r + 0.08)
        bent_y = py + lens * (1.0 if py >= 0.0 else -1.0)
        disk_y = bent_y / 0.18
        front_y = (py + 0.12) / 0.16
        back_y = (py - 0.18 - 0.50 * math.exp(-9.0 * px * px)) / 0.13
        front = accretion_band(px, front_y, 0.92, 0.035, t, density, 0.95)
        upper_arc = accretion_band(px, back_y, 0.98, 0.050, t * 0.85 + 1.7, density, 0.75)
        lower_lens = 0.42 * accretion_band(px, disk_y, 1.08, 0.060, -t * 0.55, density, 0.55)
        ring_glow = math.exp(-((r - 0.34) ** 2) / 0.005) * 0.38
        photon_ring = math.exp(-((r - 0.43) ** 2) / 0.0018) * 0.55
        shadow = smoothstep(0.41, 0.34, r)
        hole_cut = smoothstep(0.48, 0.39, r)
        doppler = 0.74 + 0.42 * smoothstep(-0.45, 0.65, px)
        disk = (front * doppler + upper_arc + lower_lens) * (1.0 - 0.70 * hole_cut)
        halo = 0.13 * math.exp(-2.5 * max(0.0, r - 0.36))
        background_stars = 0.0
        if star_noise(col + 101, row - 31) > 0.996:
          background_stars = 0.45 + 0.25 * math.sin(elapsed * 1.7 + col)
        texture = disk + photon_ring + ring_glow + halo + background_stars
        texture *= 1.0 - 0.96 * shadow
        vignette = max(0.0, 1.0 - 0.08 * abs(px) - 0.12 * abs(py))
        level = max(0.0, min(1.0, texture * vignette * options.contrast))
        char = black_hole_char(level)
        color = gray_fg(level)
        if last_color != color:
          line.append(f"\x1b[38;5;{color}m")
          last_color = color
        line.append(char)
      line.append(RESET)
      rows.append("".join(line))
    return "\n".join(rows)

File: src/animations/black_hole.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, smoothstep, FieldStyle};
use crate::noise::star_noise;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "amber" };
const TH: &[(f64, char)] = &[
  (0.10, ' '), (0.18, '.'), (0.28, ':'), (0.39, '-'), (0.51, '='),
  (0.64, '+'), (0.78, '*'), (0.91, '#'), (1.01, '%'),
];

#[inline]
fn band(x: f64, y: f64, radius: f64, thickness: f64, t: f64, density: f64, boost: f64) -> f64 {
  let dr = (x * x + y * y).sqrt();
  let dth = y.atan2(x);
  let ring = (-((dr - radius).powi(2)) / thickness).exp();
  let az = 0.62 + 0.38 * (9.0 * dth - 2.2 * t).sin();
  let turb = 0.72 + 0.28 * (31.0 * dr * density + 5.0 * dth - 3.4 * t).sin().abs();
  ring * az * turb * boost
}

pub struct BlackHole;

impl Animation for BlackHole {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width; let h = ctx.height;
    let mut grid = vec![0.0_f64; w * h];
    let radius = (w.min(h * 2) as f64 * 0.44).max(1.0);
    let density = ctx.options.scale.max(0.45);
    let t = ctx.elapsed * 0.55;
    let contrast = ctx.options.contrast;
    for row in 0..h {
      let py = ((row as f64 - (h as f64 - 1.0) * 0.5) * 2.0) / radius;
      let base = row * w;
      for col in 0..w {
        let px = (col as f64 - (w as f64 - 1.0) * 0.5) / radius;
        let r = (px * px + py * py).sqrt();
        let lens = 0.18 / (r * r + 0.08);
        let bent_y = py + lens * if py >= 0.0 { 1.0 } else { -1.0 };
        let disk_y = bent_y / 0.18;
        let front_y = (py + 0.12) / 0.16;
        let back_y = (py - 0.18 - 0.50 * (-9.0 * px * px).exp()) / 0.13;
        let front = band(px, front_y, 0.92, 0.035, t, density, 0.95);
        let upper_arc = band(px, back_y, 0.98, 0.050, t * 0.85 + 1.7, density, 0.75);
        let lower_lens = 0.42 * band(px, disk_y, 1.08, 0.060, -t * 0.55, density, 0.55);
        let ring_glow = (-((r - 0.34).powi(2)) / 0.005).exp() * 0.38;
        let photon_ring = (-((r - 0.43).powi(2)) / 0.0018).exp() * 0.55;
        let shadow = smoothstep(0.41, 0.34, r);
        let hole_cut = smoothstep(0.48, 0.39, r);
        let doppler = 0.74 + 0.42 * smoothstep(-0.45, 0.65, px);
        let disk = (front * doppler + upper_arc + lower_lens) * (1.0 - 0.70 * hole_cut);
        let halo = 0.13 * (-2.5 * (r - 0.36).max(0.0)).exp();
        let mut bg_stars = 0.0_f64;
        if star_noise(col as i64 + 101, row as i64 - 31) > 0.996 {
          bg_stars = 0.45 + 0.25 * (ctx.elapsed * 1.7 + col as f64).sin();
        }
        let mut texture = disk + photon_ring + ring_glow + halo + bg_stars;
        texture *= 1.0 - 0.96 * shadow;
        let vignette = (1.0 - 0.08 * px.abs() - 0.12 * py.abs()).max(0.0);
        grid[base + col] = clamp(texture * vignette * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
